src/DotNetty.Handlers/Timeout/WriteTimeoutHandler.cs

```diff
@@ -0,0 +1,194 @@
+    /// {
+    ///     IChannelPipeline pipeline = channel.Pipeline;
+    ///     
+    ///     pipeline.AddLast("writeTimeoutHandler", new <see cref="WriteTimeoutHandler"/>(30);
+    ///     pipeline.AddLast("myHandler", new MyHandler());
+    /// }    
+    /// </c>
+    /// 
+    /// <c>
+    /// public class MyHandler : ChannelDuplexHandler 
+    /// {
+    ///     public override void ExceptionCaught(<see cref="IChannelHandlerContext"/> context, <see cref="Exception"/> exception)
+    ///     {
+    ///         if(exception is <see cref="WriteTimeoutException"/>) 
+    ///         {
+    ///             // do somethind
+    ///         }
+    ///         else
+    ///         {
+    ///             base.ExceptionCaught(context, cause);
+    ///         }
+    ///      }
+    /// }
+    /// </c>
+    /// 
+    /// </para>
+    /// <see cref="ReadTimeoutHandler"/>
+    /// <see cref="IdleStateHandler"/>
+    /// </summary>
+    public class WriteTimeoutHandler : ChannelHandlerAdapter
+    {
+        static readonly TimeSpan MinTimeout = TimeSpan.FromMilliseconds(1);
+        readonly TimeSpan timeout;
+        bool closed;
+
+        /// <summary>
+        /// A doubly-linked list to track all WriteTimeoutTasks.
+        /// </summary>
+        readonly LinkedList<WriteTimeoutTask> tasks = new LinkedList<WriteTimeoutTask>();
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DotNetty.Handlers.Timeout.ReadTimeoutHandler"/> class.
+        /// </summary>
+        /// <param name="timeoutSeconds">Timeout in seconds.</param>
+        public WriteTimeoutHandler(int timeoutSeconds)
+            : this(TimeSpan.FromSeconds(timeoutSeconds))
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DotNetty.Handlers.Timeout.ReadTimeoutHandler"/> class.
+        /// </summary>
+        /// <param name="timeout">Timeout.</param>
+        public WriteTimeoutHandler(TimeSpan timeout)
+        {
+            this.timeout = 
+                timeout.Ticks <= 0 ? MinTimeout : TimeUtil.Max(timeout, MinTimeout);
+        }
+
+        public override Task WriteAsync(IChannelHandlerContext context, object message)
+        {
+            Task task = context.WriteAsync(message);
+
+            if (this.timeout.Ticks > 0)
+            {
+                this.ScheduleTimeout(context, task);
+            }
+
+            return task;
+        }
+
+        public override void HandlerRemoved(IChannelHandlerContext context)
+        {
+            LinkedListNode<WriteTimeoutTask> task = this.tasks.Last;
+            while(task != null)
+            {
+                task.Value.ScheduledTask.Cancel();
+                this.tasks.RemoveLast();
+                task = this.tasks.Last;
+            }
+        }
+
+        void ScheduleTimeout(IChannelHandlerContext context, Task promise)
+        {
+            var task = new WriteTimeoutTask(context, promise);
```

Comment:
Please check the latest source. I've added `OneTimeTask` so you can leverage that to better align implementation here as well.